/**
 * Copyright Soramitsu Co., Ltd. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

syntax = "proto3";
package iroha.protocol;
import "commands.proto";
import "primitive.proto";

message Transaction {
  message Payload {
    message BatchMeta{
      enum BatchType{
        ATOMIC = 0;
        ORDERED = 1;
      }
      BatchType type = 1;
      // array of reduced hashes of all txs from the batch
      repeated bytes reduced_hashes = 2;
    }
    message ReducedPayload{
      repeated Command commands = 1;
      string creator_account_id = 2;
      uint64 created_time = 3;
      uint32 quorum = 4;
    }
    // transcation fields
    ReducedPayload reduced_payload = 1;
    // batch meta fields if tx belong to any batch
    oneof optional_batch_meta{
      BatchMeta batch = 5;
    }
  }

  Payload payload = 1;
  repeated Signature signatures = 2;
}
